src/security/blocking.cpp

```diff
@@ -0,0 +1,409 @@
+    "100%;font:inherit;vertical-align:baseline}body{background:-webkit-radial-"
+    "gradient(26% 19%,circle,#fff,#f4f7f9);background:radial-gradient(circle "
+    "at 26% "
+    "19%,#fff,#f4f7f9);display:-webkit-box;display:-ms-flexbox;display:flex;-"
+    "webkit-box-pack:center;-ms-flex-pack:center;justify-content:center;-"
+    "webkit-box-align:center;-ms-flex-align:center;align-items:center;-ms-flex-"
+    "line-pack:center;align-content:center;width:100%;min-height:100vh;line-"
+    "height:1;flex-direction:column}p{display:block}main{text-align:center;"
+    "flex:1;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-"
+    "pack:center;-ms-flex-pack:center;justify-content:center;-webkit-box-align:"
+    "center;-ms-flex-align:center;align-items:center;-ms-flex-line-pack:center;"
+    "align-content:center;flex-direction:column}p{font-size:18px;line-height:"
+    "normal;color:#646464;font-family:sans-serif;font-weight:400}a{color:#"
+    "4842b7}footer{width:100%;text-align:center}footer "
+    "p{font-size:16px}</style></head><body><main><p>Sorry, you cannot access "
+    "this page. Please contact the customer service "
+    "team.</p></main><footer><p>Security provided by <a "
+    "href=\"https://www.datadoghq.com/product/security-platform/"
+    "application-security-monitoring/\" "
+    "target=\"_blank\">Datadog</a></p></footer></body></html>"sv};
+
+const std::string_view kDefaultTemplateJson{
+    "{\"errors\":[{\"title\":\"You've been blocked\",\"detail\":\"Sorry, you "
+    "cannot access this page. Please contact the customer service team. "
+    "Security provided by Datadog.\"}]}"sv};
+}  // namespace
+
+namespace datadog::nginx::security {
+
+// NOLINTNEXTLINE
+std::unique_ptr<BlockingService> BlockingService::instance;
+
+void BlockingService::initialize(std::optional<std::string_view> templ_html,
+                                 std::optional<std::string_view> templ_json) {
+  if (instance) {
+    throw std::runtime_error("Blocking service already initialized");
+  }
+  instance = std::unique_ptr<BlockingService>(
+      new BlockingService(templ_html, templ_json));
+}
+
+void BlockingService::block(BlockSpecification spec, ngx_http_request_t &req) {
+  BlockResponse const resp = BlockResponse::resolve_content_type(spec, req);
+  ngx_str_t *templ{};
+  if (resp.ct == BlockResponse::ContentType::HTML) {
+    templ = &templ_html_;
+  } else if (resp.ct == BlockResponse::ContentType::JSON) {
+    templ = &templ_json_;
+  } else {
+    req.header_only = 1;
+  }
+
+  ngx_http_discard_request_body(&req);
+
+  // TODO: clear all current headers?
+
+  req.headers_out.status = resp.status;
+  req.headers_out.content_type = BlockResponse::content_type_header(resp.ct);
+  req.headers_out.content_type_len = req.headers_out.content_type.len;
+
+  if (!resp.location.empty()) {
+    push_header(req, "Location"sv, resp.location);
+  }
+  if (templ) {
```

Comment:
I wouldn't expect there to be a template if we're redirecting (maybe an empty one?), perhaps the logic could be split to better reflect the actions taken when redirecting vs when overriding the body.

Reply:
Technically, there's nothing with sending a response entity on a redirect. In fact, it's very common. While we don't currently send anything, I don't think this method needs to reflect that.